source/git/source.go

```diff
@@ -528,49 +504,50 @@ func (gs *gitSourceHandler) Snapshot(ctx context.Context, g session.Group) (out
 		if err := os.MkdirAll(checkoutDir, 0711); err != nil {
 			return nil, err
 		}
-		_, err = gitWithinDir(ctx, checkoutDirGit, "", sock, knownHosts, nil, "-c", "init.defaultBranch=master", "init")
+		checkoutGit := git.New(gitutil.WithWorkTree(checkoutDir), gitutil.WithGitDir(checkoutDirGit))
```

Comment:
🙈

Reply:
If I'm understanding correctly, the simplest fix for #5066 is including `gitutil.WithExec(runWithStandardUmask)` here?

Reply:
... I guess all instances of `WithWorkTree` in this file technically need/should have it?

Reply:
@tianon Yes, assuming there isn't more cases like this.

But I think we need some refactoring here that would better guarantee such bugs not reappearing as described in as described in https://github.com/moby/buildkit/pull/4106#pullrequestreview-1592606139 and also make sure test can reliably cover it. cc @jedevc   We can still take your patch as a quick mitigation.

Reply:
Hm, we actually specify this option - the issue was is that we weren't correctly propagating it down to this level. Opened an alternative PR here that fixes the root cause: https://github.com/moby/buildkit/pull/5096.